File: 0038-count-and-say/0038-count-and-say.py
class Solution:
    def countAndSay(self, n: int) -> str:
        res = "1"
        for _ in range(1, n):
            res = self.build_next(res)
        return res

    def build_next(self, s: str) -> str:
        result = ""
        count = 1
        for i in range(1, len(s)):
            if s[i] == s[i - 1]:
                count += 1
            else:
                result += str(count) + s[i - 1]
                count = 1
        result += str(count) + s[-1]
        return result

File: 0038-count-and-say/0038-count-and-say.c
#include <stdlib.h>
#include <string.h>
#include <stdio.h>

char *countAndSayHelper(char *previous);

char *countAndSay(int n) {
    if (n == 1) {
        char *result = (char *)malloc(2 * sizeof(char));
        strcpy(result, "1");
        return result;
    }

    char *previous = countAndSay(n - 1);
    char *result = countAndSayHelper(previous);
    free(previous);
    return result;
}

char *countAndSayHelper(char *previous) {
    int length = strlen(previous);
    int result_capacity = length * 2 + 1;
    char *result = (char *)malloc(result_capacity * sizeof(char));
    int count = 1;
    int result_index = 0;

    for (int i = 1; i < length; ++i) {
        if (previous[i] == previous[i - 1]) {
            count++;
        } else {
            result_index += sprintf(result + result_index, "%d%c", count, previous[i - 1]);
            count = 1;
        }
    }

    result_index += sprintf(result + result_index, "%d%c", count, previous[length - 1]);
    result[result_index] = '\0';
    return result;
}
